cmd/kubectl-gadget/top/ebpf.go

```diff
@@ -0,0 +1,299 @@
+			CommonFlags:      &params,
+			Parameters: map[string]string{
+				types.MaxRowsParam:  strconv.Itoa(maxRows),
+				types.IntervalParam: strconv.Itoa(outputInterval),
+				types.SortByParam:   sortBy,
+			},
+		}
+
+		// when params.Timeout == interval it means the user
+		// only wants to run for a given amount of time and print
+		// that result.
+		singleShot := params.Timeout == outputInterval
+
+		// start print loop if this is not a "single shoot" operation
+		if singleShot {
+			ebpfPrintHeader()
+		} else {
+			ebpfStartPrintLoop()
+		}
+
+		if err := utils.RunTraceStreamCallback(config, ebpfCallback); err != nil {
+			return commonutils.WrapInErrRunGadget(err)
+		}
+
+		if singleShot {
+			ebpfPrintEvents()
+		}
+
+		return nil
+	},
+	SilenceUsage: true,
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		var err error
+		ebpfSortBy, err = types.ParseSortBy(sortBy)
+		if err != nil {
+			return commonutils.WrapInErrInvalidArg("--sort", err)
+		}
+
+		return nil
+	},
+	Args: cobra.MaximumNArgs(1),
+}
+
+func init() {
+	addTopCommand(ebpfCmd, types.MaxRowsDefault, types.SortBySlice)
+}
+
+func ebpfCallback(line string, node string) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	var event types.Event
+
+	if err := json.Unmarshal([]byte(line), &event); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %s", commonutils.WrapInErrUnmarshalOutput(err, line))
+		return
+	}
+
+	if event.Error != "" {
+		fmt.Fprintf(os.Stderr, "Error: failed on node %q: %s", event.Node, event.Error)
+		return
+	}
+
+	ebpfNodeStats[node] = event.Stats
+}
+
+func ebpfStartPrintLoop() {
+	go func() {
+		ticker := time.NewTicker(time.Duration(outputInterval) * time.Second)
+		ebpfPrintHeader()
+		for {
+			_ = <-ticker.C
+			ebpfPrintHeader()
+			ebpfPrintEvents()
+		}
+	}()
+}
+
+func ebpfPrintHeader() {
+	switch params.OutputMode {
+	case commonutils.OutputModeColumns:
+		if term.IsTerminal(int(os.Stdout.Fd())) {
+			utils.ClearScreen()
+		} else {
+			fmt.Println("")
+		}
+		fmt.Printf("%-16s %-8s %-16s %-16s %-7s %-20s %12s %8s\n",
+			"NODE",
+			"PROG",
+			"TYPE",
+			"NAME",
+			"PID", "COMM",
+			"TIME",
+			"COUNT",
+		)
+	case commonutils.OutputModeCustomColumns:
+		if term.IsTerminal(int(os.Stdout.Fd())) {
+			utils.ClearScreen()
+		} else {
+			fmt.Println("")
+		}
+		fmt.Println(ebpfGetCustomColsHeader(params.CustomColumns))
+	}
+}
+
+func ebpfPrintEvents() {
+	// sort and print events
+	mutex.Lock()
+
+	stats := []types.Stats{}
+	for node, stat := range ebpfNodeStats {
+		for i := range stat {
+			stat[i].Node = node
+		}
+		stats = append(stats, stat...)
+	}
+	ebpfNodeStats = make(map[string][]types.Stats)
+
+	mutex.Unlock()
+
+	types.SortStats(stats, ebpfSortBy)
+
+	switch params.OutputMode {
+	case commonutils.OutputModeColumns:
+		for idx, event := range stats {
+			if idx == maxRows {
+				break
+			}
+			pid := ""
+			comm := ""
+			if len(event.Pids) > 0 {
+				pid = strconv.Itoa(int(event.Pids[0].Pid))
+				comm = event.Pids[0].Comm
+			}
```

Comment:
I don't know if this has been discussed before. It seems it's too misleading to only print a pid. Would it be possible to concatenate and print them? (or at least, to have something that makes it clear to the user that not all pids were printed).

Reply:
Hmm, true. I think we could just use a new line with only the pid/comm columns filled out so that you can directly see that it refers to the line above. It'll probably not happen often anyways that a program is referenced more than once and that way we can save space (width).

Reply:
I just talked to @blanquicet about this. It would probably make sense to postpone adding the other pids until #782 is done - I'll open an issue for that, if that's okay.

Reply:
Will be handled in #793